lib/views/shared/receive.dart

```diff
@@ -175,52 +180,72 @@ class ReceiveSharedState extends ChangeNotifier {
+          currentState = ReceiveScreenStates.FileReceived;
         });
-        return client.recvFile(_code!, progress.progressHandler).then((result) {
-          result.done.then((value) async {
-            this.setState(() {
-              currentState = ReceiveScreenStates.FileReceived;
-            });
-            await tempFile.rename(nonExistingPathFor("$path" +
-                Platform.pathSeparator +
-                "${result.pendingDownload.fileName}"));
-          }, onError: defaultErrorHandler);
+        await tempFile.rename(nonExistingPathFor("$path" +
+            Platform.pathSeparator +
+            "${result.pendingDownload.fileName}"));
+      }, onError: defaultErrorHandler);
 
+      this.setState(() {
+        currentState = ReceiveScreenStates.ReceiveConfirmation;
+        acceptDownload = () {
+          controller.text = '';
+          tempFile = File(_tempPath("$path" +
+              Platform.pathSeparator +
+              "${result.pendingDownload.fileName}"));
+          var cancelFunc =
+              result.pendingDownload.accept(tempFile.writeOnlyFile());
           this.setState(() {
-            currentState = ReceiveScreenStates.ReceiveConfirmation;
-            acceptDownload = () {
-              controller.text = '';
-              tempFile = File(_tempPath("$path" +
-                  Platform.pathSeparator +
-                  "${result.pendingDownload.fileName}"));
-              var cancelFunc =
-                  result.pendingDownload.accept(tempFile.writeOnlyFile());
-              this.setState(() {
-                currentState = ReceiveScreenStates.FileReceiving;
-                this.cancelFunc = cancelFunc;
-              });
-            };
-            rejectDownload = () {
-              result.pendingDownload.reject();
-              reset();
-            };
-            pendingDownload = result.pendingDownload;
+            currentState = ReceiveScreenStates.FileReceiving;
+            this.cancelFunc = cancelFunc;
           });
+        };
+        rejectDownload = () {
+          result.pendingDownload.reject();
+          reset();
+        };
+        pendingDownload = result.pendingDownload;
+      });
+
+      return result;
+    }, onError: defaultErrorHandler);
+  }
 
-          return result;
-        }, onError: defaultErrorHandler);
+  void selectSaveDestination() async {
+    if (selectingFolder) return;
+    try {
+      this.setState(() {
+        selectingFolder = true;
+      });
+      String? directory = await FilePicker.platform
+          .getDirectoryPath(initialDirectory: prefs?.getString(PATH));
+      if (directory == null) {
+        return;
+      }
+      if (await canWriteToDirectory(directory)) {
+        setState(() {
+          saveAsPath = "$directory${Platform.pathSeparator}";
+          acceptDownload();
+        });
       } else {
         final error =
             Exception("Permission denied. Could not write to ${path!}");
```

Comment:
I am also thinking, that if we get here an error, we should try to do rejection from Receiver side, because Sender is hanging without any information, however the code is already consumed.